src/schemapack/_internals/normalize.py

```diff
@@ -123,47 +123,131 @@ def denormalize(  # noqa: PLR0912,C901
             )
             continue
 
-        if isinstance(target_ids, frozenset):
-            denormalized_object[relation_name] = []
-
-            # make the output predictable:
-            sorted_target_ids = sorted(target_ids)
-
-            for target_id in sorted_target_ids:
-                if (
-                    target_class_name in resource_blacklist
-                    and target_id in resource_blacklist[target_class_name]
-                ):
-                    raise CircularRelationError(
-                        "Cannot perform denormalization of datapack with circular relations."
-                        + " The circular relation involved the resource with id"
-                        + f" {target_id} of class {target_class_name}."
-                    )
-
-                target_resource = denormalize(
-                    datapack=datapack,
-                    schemapack=schemapack,
-                    ignored_relations=ignored_relations,
-                    _resource_blacklist=resource_blacklist,
-                    _alt_root_class_name=target_class_name,
-                    _alt_root_resource_id=target_id,
-                )
-
-                denormalized_object[relation_name].append(  # type: ignore
-                    target_resource
-                )
-
-        elif isinstance(target_ids, str):
-            denormalized_object[relation_name] = denormalize(
-                datapack=datapack,
-                schemapack=schemapack,
-                ignored_relations=ignored_relations,
-                _resource_blacklist=resource_blacklist,
-                _alt_root_class_name=target_class_name,
-                _alt_root_resource_id=target_ids,
-            )
-
-        else:
-            denormalized_object[relation_name] = None
+        next_embedding_profile = _get_next_embedding_profile(
+            embedding_profile, relation_name
+        )
+        denormalization_context = DenormalizationContext(
+            datapack=datapack,
+            schemapack=schemapack,
+            embedding_profile=next_embedding_profile,
+            resource_blacklist=resource_blacklist,
+        )
+        denormalized_object[relation_name] = _process_recursion(
+            denormalization_context,
+            target_class_name,
+            target_ids,
+        )
 
     return denormalized_object
+
+
+def _initialize_blacklist(
+    embedding_profile, root_class_name, root_resource_id, _resource_blacklist
+) -> dict[ClassName, set[ResourceId]]:
+    """Function to initialize the resource blacklist."""
+    resource_blacklist: dict[ClassName, set[ResourceId]] = defaultdict(set)
+    if embedding_profile is None:
+        resource_blacklist[root_class_name].add(root_resource_id)
+
+    if _resource_blacklist:
```

Comment:
This check could also be removed, if `_resource_blacklist` was an empty `dict` instead of `None`